lib/storage/tree.rb

```diff
@@ -0,0 +1,108 @@
+#==============================================================================
+# Copyright (C) 2022-present Alces Flight Ltd.
+#
+# This file is part of Flight Storage.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Storage is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Storage. If not, see:
+#
+# For more information on Flight Storage, please visit:
+# https://github.com/openflighthpc/flight-storage
+#==============================================================================
+require "tty-tree"
+
+module Storage
+  class Tree
+    attr_reader :name, :children
+  
+    def initialize(name, children)
+      @name = name
+      @children = [].tap do |a|
+        children.each do |child|
+          if child.is_a?(String)
+            a << child
+          elsif child.is_a?(Hash)
+            child.each do |k, v|
+              a << Tree.new(k.to_s, v)
+            end
+          end
+        end
+      end
+    end
+    
+    # Convert to a hash format that tty-tree can render properly
+    def to_hash
+      hash_children = []
+      @children.each do |child|
+        if child.class == String
+          hash_children << child
+        else
+          hash_children << child.to_hash
+        end
+      end
+      return { @name => hash_children }
+    end
+    
+    def show
+      TTY::Tree[self.to_hash].render
+    end
+
+    # Given the name of a directory in this tree, return a tree rooted at that directory
+    # Accepts as many arguments as its given, squashed into a single array named *path
+    # Recursively indexes children to find subdirectory at given path
+    def dig(*path, iteration: 0)
+      subdir = @children.find { |c| c.is_a?(Tree) && c.name == path[iter] }
+
+      if !subdir
+        # No such subdirectory
+        raise "The directory '#{File.join(*arr)}' could not be found"
+      elsif iteration + 1 == arr.length
+        # Base case
+        return subdir
+      else
+        # Continue with recursive indexing
+        iteration += 1
+        subdir.dig(*arr, iteration: iteration)
+      end
+    end
+    
+    # Given the name of a directory in this tree, return a tree rooted at that directory
+    def subtree(dir)
+      @children.each do |tree|
+        if tree.name == dir
+          return tree
+        end

```

Comment:
This still doesn't quite work. If any elements in `@children` are strings, running `tree.name` on them will raise an error, as `String` has no `name` method. I've just added a commit with a method named `Tree#dig` to find a particular directory within a tree. Could you have another look at the `file_exists?` method using that please?

Reply:
It seems like `Tree#dig`'s `arr`s are supposed to be `path`s? I assume you renamed the variable while writing the function and didn't change every instance of `arr` into `path` yet

Reply:
:grimacing: Yup, sorry, those `arr`s should definitely be `paths`.